### ISSUE TITLE
2022/02/20 - 

### ISSUE THREAD

**Issue by danbi5228** [njs03332/nlp_study#20]
![image](https://user-images.githubusercontent.com/26505830/154825845-75cc2169-1614-4553-b3e4-d9bbc0e2b13a.png)

**Comment by njs03332** [comment]
### 1.5 머신러닝의 주요 도전 과제
#### 1.5.1. 충분하지 않은 양의 훈련 데이터 
- 대부분의 머신러닝 알고리즘이 잘 작동하려면 데이터가 많아야 함
- 예) 2001년 논문: 충분한 데이터가 주어지면 자연어 중의성 해소 문제를 거의 비슷하게 잘 처리한다는 것을 입증
  - 시간과 돈을 알고리즘 개발에 쓰는 것과 말뭉치 (corpus) 개발에 쓰는 것 사이의 트레이드오프에 대해 다시 생각해봐야 한다
  - 그러나 데이터셋 작은 경우 흔하고, 훈련 데이터를 추가로 모으는 것이 항상 쉽거나 저렴한 일은 아니므로, 아직은 알고리즘을 무시하지 말아야한다

#### 1.5.2 대표성 없는 훈련 데이터
- 일반화하기 원하는 새로운 사례를 훈련 데이터가 잘 대표하는 것이 중요
- 샘플링 잡음 (sampling noise) : 우연에 의한 대표성 없는 데이터
  - 샘플이 작은 경우 발생하기 쉬움
- 샘플링 편향 (sampling bias) : 표본 추출 방법이 잘못되어 대표성을 띠지 못하는 데이터
  - 샘플이 매우 큰 경우에도 발생할 수 있음
  - 비응답 편향 (nonresponse bias) : 응답하지 않은 사람들의 특성으로 인해 생기는 편향

#### 1.5.3 낮은 품질의 데이터
- 에러, 이상치 (outlier), 잡음으로 가득한 경우 훈련 데이터 정제에 시간을 투자할만한 가치가 있음
- 훈련 데이터 정제가 필요한 경우
  - 일부 샘플이 이상치라는 게 명확한 경우
    - 해당 샘플을 무시하는 방법
    - 수동으로 고치는 방법
  - 일부 샘플에 특성 몇 개가 빠져있는 경우
    - 이 특성 모두를 무시하는 방법
    - 이 샘플을 무시하는 방법
    - 빠진 값을 채우는 방법
    - 이 특성을 넣은 모델과 제외한 모델을 따로 훈련시키는 방법

#### 1.5.4 관련 없는 특성
- 훈련 데이터에 관련 없는 특성이 적고 관련 있는 특성이 충분해야 시스템이 학습할 수 있음
- 특성 공학 (feature engineering) : 훈련에 사용할 좋은 특성들을 찾는 것
  - 특성 선택 (feature selection) : 가지고 있는 특성 중 훈련에 가장 유용한 특성을 선택
  - 특성 추출 (feature extraction) : 특성을 결합하여 더 유용한 특성을 만듦. (예: 차원 축소 알고리즘 사용)
  - 새로운 데이터를 수집해 새 특성 생성

#### 1.5.5 훈련 데이터 과적합
- 과대적합 (overfitting) : 모델이 훈련 데이터에 너무 잘 맞지만 일반성이 떨어지는 경우
- 심층 신경망 같은 복잡한 모델은 훈련 세트에 잡음이 많거나 데이터셋이 너무 작으면 잡음이 섞인 패턴을 감지하게 됨. 
  - 이런 패턴은 새로운 샘플에 일반화되지 못함
- 규제 (regularization) : 모델을 단순하게 하고 과대적합의 위험을 감소시키기 위해 모델에 제약을 가하는 것
  - 예: 선형 모델의 모델 파라미터 두 개 - 훈련 데이터에 모델을 맞추기 위한 두 개의 자유도 (degree of freedom)를 학습 알고리즘에 부여
  - 데이터에 완벽히 맞추는 것과 일반화를 위해 단순한 모델을 유지하는 것 사이의 올바른 균형을 찾는 것이 좋음
- 하이퍼파라미터 (hyperparameter) : 학습 알고리즘의 파라미터 (모델 X)
  - 학습 알고리즘으로부터 영향을 받지 않음
  - 훈련 전에 미리 지정됨
  - 훈련하는 동안에도 상수로 남아 있음
  - 머신러닝 시스템을 구축할 때 하이퍼파라미터 튜닝은 매우 중요한 과정

#### 1.5.6 훈련 데이터 과소적합
- 과소적합 (underfitting) : 모델이 너무 단순해서 데이터의 내재된 구조를 학습하지 못할 때 일어남
- 주요 해결 기법
  - 모델 파라미터가 더 많은 강력한 모델을 선택한다
  - 학습 알고리즘에 더 좋은 특성을 제공한다 (특성 공학)
  - 모델의 제약을 줄인다 (예: 규제 하이퍼파라미터를 감소)

### 1.6 테스트와 검증
- 훈련 데이터를 훈련 세트와 테스트 세트 두 개로 나누기
  - 일반화 오차 / 외부 샘플 오차 (generalization error / out-of-sample error) : 새로운 샘플에 대한 오류 비율
  - 테스트 세트에서 모델을 평가함으로써 이 와에 대한 추정값 (estimation)을 얻음
  - 이전에 본 적이 없는 새로운 샘플에 모델이 얼마나 잘 작동할지 알려줌
- 훈련 오차가 낮지만 일반화 오차가 높다면 이는 모델이 훈련 데이터에 과대적합되었다는 의미

#### 1.6.1 하이퍼파라미터 튜닝과 모델 선택
- 두 모델 중 어떤 모델을 선택할지 갈등하는 경우, 두 모델의 테스트 세트 일반화 성능을 비교하면 됨
- 시나리오) 선형 모델 사용, 하이퍼파라미터 선택을 위해 100개의 하이퍼파라미터 값으로 100개의 다른 모델을 훈련함. 그 중 일반화 오차가 가장 낮은 모델을 만드는 최적의 하이퍼파라미터를 찾아서 해당 모델을 사용함. 그러나 성능이 좋지 못한 경우
  - 일반화 오차를 테스트 세트에서 여러 번 측정하면서 모델과 하이퍼파라미터가 테스트 세트에 최적화된 모델을 만들었기 때문
- 위 문제에 대한 해결방법
  - 홀드아웃 검증 (holdout validation) : 훈련 세트의 일부를 떼어내어 여러 후보 모델을 평가하고 가장 좋은 하나를 선택
    - 검증 세트 (validation set) : 새로운 홀드아웃 세트
    - 검증 세트가 너무 작으면 모델이 정확하게 평가되지 않음
    - 검증 세트가 너무 크면 남은 훈련 세트가 전체 훈련 세트보다 너무 작아짐
  - 교차 검증 (cross-validation) : 검증 세트를 여러 개 사용해 반복적으로 검증하는 방법
    - 검증 세트마다 나머지 데이터에서 훈련한 모델을 해당 검증 세트에서 평가함
    - 모든 모델의 평가를 평균하여 정확한 성능 측정
    - 훈련 시간이 늘어난다는 단점

##### 데이터 불일치
- 시나리오) 꽃 사진에서 이름 알아내는 모델 학습
  - 웹에서 꽃 사진 다운로드 + 모바일 앱을 사용해 실제로 찍은 사진 10,000개
  - 이 경우 검증 세트와 테스트 세트가 실전에서 기대하는 데이터를 가능한 잘 대표해야 함
  - 검증 세트와 테스트 세트에 대표 사진이 배타적으로 포함되어야 함 (양쪽 세트에 비슷한 사진이 들어가지 않게)
  - 웹에서 다운로드한 사진에 모델을 훈련한 후검증 세트에서 성능 측정시 성능이 낮음
  - 모델이 훈련 세트에 과대적합되었기 때문인지, 웹 사진과 모바일 앱 사진의 데이터가 불일치하기 때문인지 알기 어려움
- 위 경우 훈련-개발 세트 (train-dev set)를 사용할 수 있음
  - 모델을 훈련한 다음 훈련-개발 세트에서 평가함
  - 모델이 잘 작동하는 경우: 데이터 불일치에서 오는 문제라는 것 -> 웹 이미지를 모바일 앱에서 찍은 사진처럼 보이도록 전처리한 다음 다시 훈련
  - 모델이 훈련-개발 세트에서 잘 작동하지 않는 경우: 훈련 세트에 과대적합된 것 -> 모델을 규제하거나 더 많은 훈련 데이터를 모으거나 훈련 데이터 정제 시도

**Comment by givitallugot** [comment]
## 1.1 머신러닝이란?

- 머신러닝: 데이터에서부터 학습하도록 컴퓨터를 프로그래밍하는 과학(또는 예술)
  - training set: 훈련 데이터
  - accuracy: 분류의 성능 측정을 위한 지표
 
- 참고로,
  - 통계: 자료를 잘 성명하는 모형을 찾는 것까지
  - 머신러닝: 학습할 수 있는 모형을 만들어주고, 자료가 들어가면 학습이 계속 업데이트되는 것까지 (머신이 알아서 러닝을 하는 관점)

## 1.2 왜 머신러닝을 사용하는가?

1. 유지 보수가 쉬움, 정확도 높은 모델
2. 알고리즘이 학습한 것을 조사하여 연관 관계나 새로운 추세 발견 (**데이터 마이닝**)

- 전통적인 프로그래밍: 스팸에 어떤 단어들이 나타나는지 -> 규칙 발견 -> 스팸으로 분류 / **유지 보수가 힘듦**
- 머신러닝 기반: 패턴을 감지하여 어떤 단어와 구절이 스팸을 판단하는 데 좋은 기준인지 자동으로 학습 / **프로그램이 짧아지고 유지 보수가 쉬움, 정확도 높음**

#### 머신러닝이 뛰어난 분야?
- 기존 솔루션으로는 많은 수동 조정과 규칙이 필요한 문제
- 전통적인 방식으로는 해결 방법이 없는 복잡한 문제
- 유동적인 환경 (새로운 데이터 적응)
- 복잡한 문제와 대량의 데이터에서 통찰 얻기

**Comment by danbi5228** [comment]
## 1.3 애플리케이션 사례
- 제품이미지를 분석해 자동으로 분류하기; CNN
- 뇌 스캔을 통해 종양 진단하기; CNN
- 자동으로 뉴스 기사 분류하기; CNN. 텍스트 분류
- 부정적 코멘트 분류하기; 텍스트 분류. NLP 도구사용
- 다양한 지표를 기반으로 회사의 내년도 수익 예측; RNN, CNN, Transformer. 회귀작업
- 구매 이력 기반으로 고객 분류; clustering
- 고차원의 데이터셋을 명확하고 의미있는 그래프로 표현; 차원 축소
- 과거 구매 이력을 기반으로한 상품 추천; 인공 신경망에 주입하여 가능성이 가장 높은 상품 출력. 추천 시스템
- 알파고와 같은 지능형 로봇; 강화학습

## 1.4 머신러닝 시스템의 종류

- 머신러닝 구분 방법
   - 사람의 감독하에 훈련하는지 여부 (지도, 비지도, 준지도, 강화 학습)
   - 점진적 학습인지, 전체 1회 학습인지 (온라인, 배치 학습)
   - 기존 데이터 포인트와 비교하는지, 패턴을 발견하는 것인지 (사례 기반, 모델 기반 학습)
- 위 구분 방법은 서로 배타적이지 않고, 원하는대로 연결 가능
   - ex) 심층 신경망 모델을 사용해서 스팸과 스팸이 아닌 메일을 가지고 실시간으로 학습한 스팸 필터기능
   - 온라인, 모델기반, 지도학습 시스템

### 1.4.1 지도 학습 vs 비지도 학습
- 지도 학습
  - 훈련 데이터에 레이블이라는 원하는 답이 포함됨
  - 분류(classfication) 작업과 회귀 작업(regression. 예측 변수라 부르는 특성(Feature)을 사용해 타깃(Target) 수치를 예측하는 작업) 이 전형적인 지도 학습
  - 알고리즘 예시
     - K-최근접 이웃
     - 선형 회귀
     - 로지스틱 회귀
     - 서포트 벡터 머신
     - 결정 트리와 랜덤 포레스트
     - 신경망
- 비지도 학습
   - 훈련 데이터에 레이블이 없음. 아무런 도움 없이 학습
   - 아래 작업들이 비지도 학습에 포함되는 중요한 작업
   - 계층 군집 알고리즘 : 각 그룹을 더 작은 그룹으로 세분화
   - 시각화 알고리즘 : 데이터 조직 이해 및 예상치 못한 패턴 발견 가능
   - 차원 축소 : 여러 특성을 하나로 합치는 것. 특성 추출이 가능 (차원축소를 바탕으로 하나의 특성으로 추출하는 것)
   - 이상치 탐지 : 정상 데이터와 다른 데이터 탐지
      - cf. 특이치 탐지 : 훈련 세트에 있는 모든 샘플과 달라보이는 새로운 샘플 탐지
      - ex) 99% 강아지 사진 + 1% 치와와 사진 : 이상치 탐지의 경우 치와와 사진 탐지O. 특이치 탐지의 경우 치와와 사진 탐지X
   - 연관 규칙 학습 : 대량의 데이터에서 특성 간의 흥미로운 관계를 찾음
   - 알고리즘 예시
      - 군집 clustering
         - k-평균
         - DBSCAN
         - 계층 군집 분석
         - 이상치 탐지와 특이치 탐지
         - 원-클래스 SVM
         - isolation forest
      - 시각화와 차원축소
         - PCA
         - 커널 PCA
         - 지역적 선형 임베딩
         - t-SNE
      - 연관 규칙 학습
         - 어프라이어리
         - 이클렛 
- 준지도 학습
   - 일부만 레이블이 있는 데이터를 다루는 경우
   - 대부분 지도학습과 비지도 학습의 조합
- 강화 학습
   - 에이전트 : 학습하는 시스템
   - 환경을 관찰해서 행동을 실행하고 그 결과로 보상/벌점을 받고, 가장 큰 보상을 얻기 위한 정책을 스스로 학습
   - 정책은 주어진 상황에서 에이전트가 어떤 행동을 선택해야 할지 정의

### 1.4.2 배치 학습과 온라인 학습 
- 배치 학습
   - 점진적으로 학습할 수 없음. 가용한 데이터를 모두 사용해 훈련시킴
   - 오프라인 학습 : 학습한 것을 단지 적용만 하고 더이상의 학습 없이 실행됨
   - 새로운 데이터에 대해 학습하기 위해서는 이전 데이터를 포함해서 새로운 버전으로 처음부터 다시 훈련 후, 완료되면 시스템 교체
   - 장점: 간단하고 잘 동작함
   - 단점: 컴퓨팅 자원이 많이 필요함. 훈련에 많은 시간이 필요할 수 있음. 데이터 양이 너무 많으면 배치 학습 자체가 불가능
- 온라인 학습
   - 데이터를 순차적으로 한 개씩 혹은 미니배치 라는 작은 묶음 단위로 주입하여 시스템을 훈련시킴
   - 학습률: 온라인 학습에 필요한 파라미터. 변화하는 데이터에 얼마나 빠르게 적응할 것인지에 대한 수치
      - 높게하면 데이터에는 빠르게 적응하나 이전 데이터를 잊어버릴 것
      - 낮게하면 새로운 데이터에 대한 적응은 느리지만 잡음에 덜 민감할 것
   - 장점: 매 학습 단계가 빠름. 비용이 적게 듦

### 1.4.3 사례 기반 학습과 모델 기반 학습
- 사례 기반 학습
   - 시스템이 훈련 샘플을 기억하여 학습. 유사도 측정을 사용해 새로운 데이터와 학습한 샘플을 비교하는 식으로 일반화
- 모델 기반 학습
   - 샘플들의 모델을 만들어 예측에 사용하는 것
   - 효용함수 (모델이 얼마나 좋은지 측정)와 비용함수(모델이 얼마나 나쁜지 측정)를 정의하여 모델 성능 측정 지표를 정할 수 있음
   - 선형 회귀에서는 예측화 훈련 데이터 사이의 거리를 재는 비용 함수를 사용
      - 알고리즘에 데이터를 공급하면 가장 잘 맞는 선형 모델의 파라미터를 찾아냄 (훈련)
      - 이를 기반으로 기존에 없는 데이터에 대해서, 파라미터 값을 가지고 결과를 낼 수 있음 (p.51 참고)
